Add API for resetting state of a `SystemIndexPlugin`

williamrandolph opened the pull request:
When we disable access to system indices, plugins will still need a way to erase their state. The obvious and most pressing use case for this is in tests, which need to be able to clean up the state of a cluster in between groups of tests.

In this first draft of the feature state reset API, a user with snapshot admin privileges `POST`s a request to `/_reset_feature_state`. The node handling the endpoint then iterates through a list of features and executes their cleanup callbacks. By default, these callbacks use a client and a clusterState to delete all system indexes and associated indexes for a feature. The endpoint then returns a list of objects that include a feature name and a report of whether the reset operation succeeded or failed.

williamrandolph committed:
Stub out classes for reset feature states API

williamrandolph committed:
Add enough code to get an integration test working

williamrandolph committed:
Improve unit test

williamrandolph committed:
Add rough version of callback to SystemIndexPlugin

williamrandolph committed:
License changes

williamrandolph committed:
Use a HandledTransportAction for reset action

My initial cut used a TransportMasterNodeAction, which requires code
that carefully manipulates cluster state. At least for the first cut and
testing, it seems like it will be much easier to use a client within a
HandledTransportAction, which effectively makes the
TransportResetFeatureStateAction a class that dispatches other transport
actions to do the real work. This could leave cleanup vulnerable to
interference from other actions, e.g., indices deleted between lookup
and deletion actions, but I can try to tighten that up at a later stage
if I need to.

williamrandolph committed:
Just use clusterService, client, and listener in callback

williamrandolph committed:
Add javadoc and update tests

williamrandolph committed:
Fill out ResetFeatureStateResponse class

williamrandolph committed:
Hook up response class in TransportAction

williamrandolph committed:
Rename inner class

williamrandolph committed:
Add a second plugin to integration tests

williamrandolph committed:
Clean up code by using a GroupedActionListener

williamrandolph committed:
Merge branch 'master' into si/reset-features-api

williamrandolph commented:
I still have some TODOs scattered in the code. I'll keep working on those, but I'm interested in feedback on the overall approach.

williamrandolph committed:
Add reset to list of non-operator actions

AthenaEryma committed:
ML feature state cleaner

gwbrown commented on server/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestResetFeatureStateAction.java:
```diff
@@ -0,0 +1,41 @@
+        final ResetFeatureStateRequest req = new ResetFeatureStateRequest();
+
+        return restChannel -> {
+            client.execute(SnapshottableFeaturesAction.INSTANCE, req, new RestToXContentListener<>(restChannel));
```
```suggestion
            client.execute(ResetFeatureStateAction.INSTANCE, req, new RestToXContentListener<>(restChannel));
```

As written, this explodes if you try to call this API via the REST interface. Easy enough copy/paste error.

AthenaEryma reviewed (commented):
Left a couple comments, approach looks good though. Nice work!

williamrandolph committed:
Consolidate default deletion logic in SystemIndices.feature

AthenaEryma committed:
Implement Transform feature state reset

williamrandolph committed:
Merge changes from master

williamrandolph commented on server/src/main/java/org/elasticsearch/plugins/SystemIndexPlugin.java:
```diff
@@ -49,4 +60,40 @@
+     * @param client A client, for executing actions
+     * @param listener Listener for post-cleanup result TODO[wrb]: need to gather results over features
+     */
+    default void cleanUpFeature(
```
I added a check for whether or not there are indices to delete before issuing the request, but also moved the logic to `SystemIndices.Feature#cleanUpFeature` to make it available to built-in features rather than just plugins.

williamrandolph committed:
Merge changes from master

williamrandolph committed:
Change endpoint to _feature/reset

williamrandolph committed:
Add javadoc and cleanup java style

AthenaEryma committed:
Merge branch 'si/reset-transforms' into si/reset-features-api

hendrikmuhs commented on x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/Transform.java:
```diff
@@ -377,6 +380,35 @@ public void close() {
         return List.of(AUDIT_INDEX_PATTERN);
```
@gwbrown 

- Is this + the `getSystemIndexDescriptors` the list of indexes to be deleted at cleanup?
- Is this used _only_ for deletion or for something else, too?

There is 1 case to be aware of, if a user started with an older version, used transform on it and upgraded the installation once or more, he has older indices still lying around. Those must be deleted at reset as well. 

It seems to me, this is an overlook introduced earlier(this PR is not to blame). Let me know, I can create a tiny separate PR once I have clarification from you.

With other words this should return `AUDIT_INDEX_PATTERN + INDEX_NAME_PATTERN_DEPRECATED + .transform-internal-00{1-5}` (with `.transform-internal-006` we switched over to a system index)

gwbrown replied:
1. Yes, this and `getSystemIndexDescriptors` are used for deletion.
2. Associated indices are also automatically included in a snapshot if the plugin's feature state (that is, system indices) are included in the snapshot, to ensure the snapshot can be used to restore all data owned by the feature. That's the only other use of associated indices at this time, though (and we don't have immediate plans for more, but there's always possibilities).

It sounds like yes, we should include the deprecated patterns in the associated indices. The system index descriptor already uses the `.transform-internal-*` indices:
https://github.com/elastic/elasticsearch/blob/54ed2e37d9fc5432c9c76ab05cded8d11a64133e/x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/persistence/TransformInternalIndex.java#L90
So those will be picked up as part of the system indices - they don't need to (and shouldn't) be registered as associated indices as well.

hendrikmuhs replied:
ok, thanks. That means the only uncovered one is `TransformInternalIndexConstants.INDEX_NAME_PATTERN_DEPRECATED` (needed only for the `7.x` series, for `8` we plan to provide an upgrader that gets rid of the deprecated indexes)

williamrandolph committed:
Add stub for API definition

williamrandolph committed:
Merge branch 'si/reset-features-api' of github.com:williamrandolph/elasticsearch into si/reset-features-api

williamrandolph committed:
Add stub for RHLC feature resetting

williamrandolph committed:
Implement client feature reset response parser

williamrandolph committed:
Merge branch 'master' into si/reset-features-api

williamrandolph committed:
Merge branch 'master' into si/reset-features-api

williamrandolph committed:
Respond to PR feedback

williamrandolph committed:
Change _features/reset path to _features/_reset

Out of an abundance of caution, I think the "reset" part of this path
should have a leading underscore, so that if there's ever a reason to
implement "GET _features/<feature_id>" we won't have to worry about
distinguishing "reset" from a feature name.

williamrandolph committed:
Checkstyle fixes

williamrandolph commented on server/src/main/java/org/elasticsearch/indices/SystemIndices.java:
```diff
@@ -269,5 +314,56 @@ public String getDescription() {
+            List<String> associatedIndices = new ArrayList<>(associatedIndexPatterns);
+
+            List<String> allIndices = Stream.concat(systemIndices.stream(), associatedIndices.stream())
+                .collect(Collectors.toList());
```
I made the suggested change, but you're probably right about associatedIndexPatterns. I'll take a look and report back.

williamrandolph replied:
This works as written, because we can pass a valid index pattern to the delete request, which will then handle the resolution logic when executed. We can't do the same thing with system index descriptor patterns because they can use character-class regexes, for example `.test-[ab]*`.

It might be worth resolving the associated indices here so that we can skip submitting the delete request when no indices exist. Do you think this is important to do?

jaymode replied:
I think it is fine as it is unless we discover issues

williamrandolph commented on server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/features/ResetFeatureStateRequest.java:
```diff
@@ -0,0 +1,39 @@
+public class ResetFeatureStateRequest extends ActionRequest {
+
+    public ResetFeatureStateRequest() {
+        // TODO[wrb] - We might need to let this request take a list of
```
Created #70426

jaymode reviewed (approved):
LGTM

williamrandolph commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into si/reset-features-api

williamrandolph commented:
@elasticmachine run elasticsearch-ci/1

elasticmachine commented:
Pinging @elastic/es-core-infra (Team:Core/Infra)

williamrandolph commented:
Backport to 7.x: https://github.com/elastic/elasticsearch/pull/70524
